// ngx-translate-cache/esm5/lib/ngx-translate-cache.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CACHE_NAME, CACHE_MECHANISM, COOKIE_EXPIRY, TranslateCacheSettings } from './ngx-translate-cache.service';
var TranslateCacheModule = /** @class */ (function () {
    function TranslateCacheModule() {
    }
    TranslateCacheModule_1 = TranslateCacheModule;
    TranslateCacheModule.forRoot = function (config) {
        return {
            ngModule: TranslateCacheModule_1,
            providers: [
                { provide: CACHE_NAME, useValue: config.cacheName },
                { provide: CACHE_MECHANISM, useValue: config.cacheMechanism },
                { provide: COOKIE_EXPIRY, useValue: config.cookieExpiry },
                TranslateCacheSettings,
                config.cacheService,
            ]
        };
    };
    var TranslateCacheModule_1;
    TranslateCacheModule = TranslateCacheModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule
            ],
            declarations: [],
            exports: []
        })
    ], TranslateCacheModule);
    return TranslateCacheModule;
}());
export { TranslateCacheModule };
