// ngx-translate-cache/esm2015/lib/ngx-translate-cache.module.js
var TranslateCacheModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CACHE_NAME, CACHE_MECHANISM, COOKIE_EXPIRY, TranslateCacheSettings } from './ngx-translate-cache.service';
let TranslateCacheModule = TranslateCacheModule_1 = class TranslateCacheModule {
    static forRoot(config) {
        return {
            ngModule: TranslateCacheModule_1,
            providers: [
                { provide: CACHE_NAME, useValue: config.cacheName },
                { provide: CACHE_MECHANISM, useValue: config.cacheMechanism },
                { provide: COOKIE_EXPIRY, useValue: config.cookieExpiry },
                TranslateCacheSettings,
                config.cacheService,
            ]
        };
    }
};
TranslateCacheModule = TranslateCacheModule_1 = __decorate([
    NgModule({
        imports: [
            CommonModule
        ],
        declarations: [],
        exports: []
    })
], TranslateCacheModule);
export { TranslateCacheModule };
